src/modular-etherspot-wallet/wallet/ModularEtherspotWallet.sol

```diff
@@ -0,0 +1,248 @@
+        emit ModuleUninstalled(moduleType, module);
+    }
+
+    /**
+     * Validator selection / encoding is NOT in scope of this standard.
+     * @inheritdoc IERC7579Account
+     */
+    function validateUserOp(
+        PackedUserOperation calldata userOp,
+        bytes32 userOpHash,
+        uint256 missingAccountFunds
+    )
+        external
+        payable
+        virtual
+        override
+        payPrefund(missingAccountFunds)
+        returns (uint256 validSignature)
+    {
+        address validator;
+        // @notice validator encodig in nonce is just an example!
+        // @notice this is not part of the standard!
+        // Account Vendors may choose any other way to impolement validator selection
+        uint256 nonce = userOp.nonce;
+        assembly {
+            validator := shr(96, nonce)
+        }
+
+        // check if validator is enabled. If terminate the validation phase.
+        if (!_isValidatorInstalled(validator)) return VALIDATION_FAILED;
+
+        // bubble up the return value of the validator module
+        validSignature = IValidator(validator).validateUserOp(
+            userOp,
+            userOpHash
+        );
+    }
+
+    /**
+     * @dev ERC-1271 isValidSignature
+     *         This function is intended to be used to validate a smart account signature
+     * and may forward the call to a validator module
+     *
+     * @param hash The hash of the data that is signed
+     * @param data The data that is signed
+     */
+    function isValidSignature(
+        bytes32 hash,
+        bytes calldata data
+    ) external payable virtual override returns (bytes4) {
+        address validator = address(bytes20(data[0:20]));
+        if (!_isValidatorInstalled(validator)) revert InvalidModule(validator);
+        return
+            IValidator(validator).isValidSignatureWithSender(
+                msg.sender,
+                hash,
+                data[20:]
+            );
+    }
+
+    /**
+     * @inheritdoc IERC7579Account
+     * @param additionalContext is not needed here. It is only used in cases where the modules are
+     * stored in more complex mappings
+     */
+    function isModuleInstalled(
+        uint256 moduleType,
+        address module,
+        bytes calldata additionalContext
+    ) external view override returns (bool isInstalled) {
+        if (moduleType == MODULE_TYPE_VALIDATOR)
+            return _isValidatorInstalled(module);
+        else if (moduleType == MODULE_TYPE_EXECUTOR)
+            return _isExecutorInstalled(module);
+        else if (moduleType == MODULE_TYPE_FALLBACK)
+            return _isFallbackHandlerInstalled(module);
+        else return false;
+    }
+
+    /**
+     * @inheritdoc IERC7579Account
+     */
+    function accountId()
+        external
+        view
+        virtual
+        override
+        returns (string memory)
+    {
+        // vendor.flavour.semver
+        return "uMSA.simple/noHook.v0.1";
```

Comment:
Shouldn't this read "Etherspot/v0.1"?

Reply:
Sorry yes, amended to "ModularEtherspotWallet/.v0.1"